LPCXpresso55S69 board

agansari opened the pull request:
Adding LPC55S69 port to Zephyr RTOS.
 - mcu initialization
 - uart support
 - gpio pins and leds
 - board documentation
 - how to setup jlink and daplink(flash copy only)

Fixes #15647

agansari committed:
soc: added LPC5500 family

SoC implementation for LPC5500 family.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
dts: LPC55S69 initial device tree

Create device tree for LPCXpresso55S69 board.
Single core (cpu0) is currently supported.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
doc: lpclink2 firmware reference

Documentation updated, added reference to the latest
LPClink2 firmware, default provided firmware may not
work on latest boards (e.g. LPC55xx).

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

galak commented on boards/arm/lpcxpresso55S69/board.cmake:
```diff
@@ -0,0 +1,24 @@
+endif()
+
+if(CONFIG_BOARD_LPCXPRESSO55S69_CPU0)
+board_runner_args(jlink "--device=LPC55S69")
```
I found that 'LPC55S69_core0' and 'LPC55S69_core1' I think work for jlink.

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+board.
+
+:ref: `opensda-daplink-onboard-debug-probe`
+-----------------------------------------
```
Section heading underline needs to be at least as long as the title (add some more ``---``)

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+.. _LPCXPRESSO55S69 Schematics:
+   https://www.nxp.com/webapp/Download?colCode=LPC55S69-SCH
+
+.. _updating LPCXpresso firmware
```
add a colon at the end:  ``.. _updating LPCXpresso firmware:``

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
++---------+-----------------+----------------------------+
+| PIO1_6  | GPIO            | BLUE_LED                   |
++---------+-----------------+----------------------------+
+| PIO1_7_ | GPIO            | GREEN LED                  |
```
If you need to have an underscore here, write this as ``PIO1_7\_`` (same on line 88 above).  Otherwise delete both of these trailing underscores.

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+   ***** Booting Zephyr OS zephyr-v1.14.0 *****
+   Hello World! lpcxpresso55S69_cpu0
+
+.. _LPC55S69 SoC Website:
```
Need to also add a link for ``updating LPCXpresso firmware`` you're referencing on line 137 above.

agansari replied:
No need, the default is now JLink. DapLink is not yet fully working, the tutorial is upgrading the debug firmware is there to let you know you have other options, but tools are under development.

I added "This example uses the :ref:`jlink-debug-host-tools` as default. to both flashing and debugging". What do you think?

dbkinder commented on doc/guides/debugging/probes.rst:
```diff
@@ -55,6 +55,8 @@ This probe is realized by programming the LPC-Link2 microcontroller with J-Link
 LPC-Link2 firmware. Download and install `LPCScrypt`_ to get the firmware and
 programming scripts.
 
+**Please make sure the firmware supports the board by visiting `Firmware for LPCXpresso`_**
```
This would render better as:
```
.. note:: Verify the firmware supports your board by visiting `Firmware for LPCXpresso`_
```

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+subsystem and accelerometer, with several options for adding off-the-shelf add-on
+boards for networking, sensors, displays and other interfaces.
+
+.. image:: ./lpcxpresso55s69.png
```
rename the ``lpc55s69_evk.png`` image included with this PR to ``lpcxpresso55s69.png`` (our convention is to have the board, board doc, and board image name be the same).

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+The LPCXpresso55S69 development board provides the ideal platform for evaluation
+of and development with the LPC55S6x MCU based on the Arm® Cortex®-M33
+architecture. The board includes a high performance onboard debug probe, audio
+subsystem and accelerometer, with several options for adding off-the-shelf add-on
```
```suggestion
subsystem, and accelerometer, with several options for adding off-the-shelf add-on
```

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+of and development with the LPC55S6x MCU based on the Arm® Cortex®-M33
+architecture. The board includes a high performance onboard debug probe, audio
+subsystem and accelerometer, with several options for adding off-the-shelf add-on
+boards for networking, sensors, displays and other interfaces.
```
```suggestion
boards for networking, sensors, displays, and other interfaces.
```

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+- LPCXpresso-V3 expansion option compatible with Arduino UNO
+- PMod compatible expansion / host connector
+
+For more information about the LPC55S69 SoC and LPCXPRESSO55S69 board:
```
```suggestion
For more information about the LPC55S69 SoC and LPCXPRESSO55S69 board, see:
```

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
++-----------+------------+-------------------------------------+
+
+The default configuration only enables the first core for now and can
+be found in the following defconfig file:
```
Change to:
```
The default configuration file
``boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0_defconfig``
only enables the first core.
```

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+:ref: `opensda-daplink-onboard-debug-probe`
+-----------------------------------------
+
+PyOCD support for this board is onging and not yet available.
```
```suggestion
PyOCD support for this board is ongoing and not yet available.
```

dbkinder commented on boards/arm/lpcxpresso55S69/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+Regardless of your choice in debug probe, we will use the LPC-Link2
+microcontroller as a usb-to-serial adapter for the serial console.
+
+Connect a USB cable from your PC to P6
```
Let's merge these two sentences into one "step":
```
Connect a USB cable from your PC to P6, and use the
serial terminal of your choice (minicom, putty, etc.)
with the following settings:
```

dbkinder reviewed (changes requested):
Missed clicking on Request changes :)

ioannisg commented on soc/arm/nxp_lpc/lpc55xxx/Kconfig.series:
```diff
@@ -0,0 +1,14 @@
+	bool "LPC5500 Series Family MCU"
+	select HAS_MCUX
+	select SOC_FAMILY_LPC
+	select CPU_HAS_SYSTICK
```
If the series only include CM33 variants, CPU_HAS_SYSTICK is not needed. Actually the option is now called CPU_CORTEX_M_HAS_SYSTICK

ioannisg commented on soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc:
```diff
@@ -0,0 +1,45 @@
+
+config SOC_LPC55S69_CPU0
+	bool "SOC_LPC55S69 M33 [CPU 0]"
+	select CPU_CORTEX_M33
```
Could this selection be moved in the _SERIES? It looks common to the tw CPUs.

ioannisg reviewed (changes requested):
Looks ok, I left a couple of comments.

agansari commented:
Thank you guys @MaureenHelm @galak @ioannisg @dbkinder made the changes you required. Can you have another look?

MaureenHelm commented on soc/arm/nxp_lpc/Kconfig:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2017, NXP
+# Copyright (c) 2019, NXP
```
Don't remove existing copyrights. I meant in https://github.com/zephyrproject-rtos/zephyr/pull/15647#discussion_r279887544 to update the copyright in new files.

agansari replied:
@MaureenHelm made the changes you required

MaureenHelm reviewed (changes requested):
Please make the `fsl_{power,reset}` changes for lpc54114 and update `ext/hal/nxp/mcux/scripts/import_mcux_sdk.py` in a separate commit before you import the sdk for lpc55s69.

agansari committed:
ext: LPC54114 fsl_power and fsl_reset moved to device

Since LPC devices use different power hardware, a separation is
required between device specific drivers and common drivers.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
ext: import_mcux_sdk.py script changes

Importing MCUX drivers will copy fsl_power and fsl_reset to devices,
not drivers as before. Script will exclude 'freertos' drivers from SDK.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
boards: lpcxpresso55s69 initalization

Board LPCXpresso55S69 added to supported zephyr boards, initial simple
configuration to boot board and use UART.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

MaureenHelm reviewed (changes requested):
Looks much better, thank you. Please start working to resolve the CI failures, except for the license check which is expected to fail due to new files being imported into `ext/`.

agansari commented on ext/hal/nxp/mcux/devices/LPC55S69/fsl_power.c:
```diff
@@ -3,7 +3,6 @@
  * Copyright (c) 2016, NXP
  * All rights reserved.
  *
- *
```
Yes, I see this now, didn't pay attention when resetting the files. I used git mv, but as move is in one commit and delete in another, the files were duplicated. Need to do another forcepush besides the one below.

Also I think I found the issue with failed CI, it was related to CMakeLists.txt, couldn't build other mcux devices as fsl_reset and fsl_power were expected in device folder for Kinetis+i.MX as well.

agansari replied:
@MaureenHelm fixed all the bad file versioning issues in my former commit and automated checks now pass, except for licence.

ioannisg replied:
This looks merge-able now, IMHO

MaureenHelm commented on ext/hal/nxp/mcux/devices/LPC55S69/fsl_powerquad_basic.c:
```diff
@@ -0,0 +1,126 @@
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "fsl_powerquad.h"
```
powerquad should go in the shared drivers folder instead

agansari replied:
Thank you @MaureenHelm powerquad was filtered as part of fsl_power by the script. Waiting for a review from @dbkinder and @MeganHansen for another force push.

MaureenHelm reviewed (changes requested):
@dbkinder please revisit

agansari committed:
ext: update nxp mcux lpc sdk

Update to latest MCUXpresso SDK 2.5.1 for LPC family.
Added LPC55S69 device.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

galak reviewed (dismissed):
some minor cleanups.

galak commented on dts/arm/nxp/nxp_lpc55S6x.dtsi:
```diff
@@ -0,0 +1,88 @@
+			reg = <0x10000000 DT_SIZE_K(630)>;
+		};
+
+		usart0:usart@50086000 {
```
we need to look at structuring this similar to how dts/arm/nordic/nrf9160* is done.  So that secure/non-secure can work cleanly in the future.

agansari replied:
@galak i would make a separate issue from this pull request to separate secure and non-secure configuration. I also think that it can be done different from nrf9160 because sec/nonsec HW addresses for memory and peripherals are offseted by 0x10000000.

galak replied:
Ok, can we at least use the same address range for all the peripherals?

galak commented on dts/arm/nxp/nxp_lpc55S6x.dtsi:
```diff
@@ -0,0 +1,88 @@
+
+		gpio0: gpio@0 {
+			compatible = "nxp,kinetis-gpio";
+			reg = <0x4008c000 0x2488>;
```
Can we make this `0x5008c000` instead?

galak commented on dts/arm/nxp/nxp_lpc55S6x.dtsi:
```diff
@@ -0,0 +1,88 @@
+
+		gpio1: gpio@1 {
+			compatible = "nxp,kinetis-gpio";
+			reg = <0x4008c000 0x2488>;
```
Can we make this `0x5008c000`

agansari replied:
@galak moved all memory and peripherals to secure address space, validated and force pushed.

Also created #16330 to have secure/non-secure split.

dbkinder reviewed (approved):
doc change LGTM, thanks.

MaureenHelm commented on ext/hal/nxp/mcux/README:
```diff
@@ -17,6 +17,7 @@ Status:
    ----------------------------------------------------------------------------
    MIMXRT1015	SDK 2.5.0 (2019-03-11)		REL_2.5.0_RT1015_RFP
    LPC54114	SDK 2.5.0 (2018-12-17)		REL_2.5.0_REL9_RFP_RC3_7_1
+   LPC55S69	SDK 2.5.1 (2019-04-18)		REL_SDK_NIOBE4_2.5.1_RFP_RC1
```
Per https://github.com/zephyrproject-rtos/zephyr/pull/16049#discussion_r283954331, please use the "revision" attribute instead of the tag.

agansari replied:
@MaureenHelm done!

MaureenHelm reviewed (changes requested):
Just one nit, otherwise LGTM
